src/librustdoc/clean/types.rs

```diff
@@ -980,6 +969,38 @@ impl AttributesExt for [ast::Attribute] {
     }
 }
 
+impl AttributesExt for [ast::Attribute] {
+    type AttributeIterator<'a> = impl Iterator<Item = ast::NestedMetaItem> + 'a;
+    type Attributes<'a> = impl Iterator<Item = &'a ast::Attribute> + 'a;
+
+    fn lists<'a>(&'a self, name: Symbol) -> Self::AttributeIterator<'a> {
+        self.iter()
+            .filter(move |attr| attr.has_name(name))
+            .filter_map(ast::Attribute::meta_item_list)
+            .flatten()
+    }
+
+    fn iter<'a>(&'a self) -> Self::Attributes<'a> {
+        self.into_iter()
+    }
+}
+
+impl AttributesExt for [(ast::Attribute, Option<DefId>)] {
```

Comment:
This impl forces attribute cloning everywhere.
See how `Attributes::from_ast_iter` is implemented - it uses `(&'a ast::Attribute, Option<DefId>)` to use attribute references pointing to tcx arenas without cloning.

(Also, I'm not sure this `AttributesExt` trait is useful in general, it's never used generically, only for providing method syntax.)

Reply:
Good points. I'll see how to improve all that.